fix(databases): GSheets and Clickhouse DBs are not allowed to upload files

Antonio-RiveroMartnez opened the pull request:
### SUMMARY
Our users are getting an `Internal Server Error 500` when trying to upload files to `GSheets` databases, that's because such databases don't support file upload nor the `Clickhouse` ones thus we need to stop showing such options to the users and prevent that existing ones get displayed in the upload form.

Additionally,  `schemas allowed` field is dependent on having file upload enabled but doesn't appear that way, so we are rendering it only when user checks the `Allow file upload` checkbox and placing it underneath the file upload checkbox.

Wording changes:
1. Update field name to : "Allow file uploads to database" instead of "Allow data upload" and remove the tooltip
2. On the Schemas Allowed field, we should update "CSV Upload" text to be "File Upload" and "CSVs" to "files", including in the hint text


### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
Before:
![error](https://user-images.githubusercontent.com/38889534/184278578-5969753c-d5a3-4ff3-9895-130cfd51d975.gif)

![error](https://user-images.githubusercontent.com/38889534/184930260-5d380b6a-fb6e-4cd9-b419-c11bba03878a.gif)


After:
![test](https://user-images.githubusercontent.com/38889534/184278136-0788af68-ce22-484a-bc2f-3608517c2b80.gif)

![test](https://user-images.githubusercontent.com/38889534/184930361-9b8230f6-9d52-47ff-aed6-8169c6825f2e.gif)



### TESTING INSTRUCTIONS
1. When selecting a DB type, if GSheet or Clickhouse is selected the `Allow data upload` checkbox is not displayed in the form
2. Enabled the file upload menu only if there is any exiting DB with allow_file_upload flag set to true and that is not `GSheets` nor `Clickhouse`
3. When the form to file upload files is opened, the `Database` dropdown must not contain any `GSheets` nor `Clickhouse` DB.

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [X] Has associated issue: https://github.com/apache/superset/issues/19247
- [ ] Required feature flags:
- [X] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

Antonio-RiveroMartnez committed:
GSheets and Clickhouse Databases:

- GSheets  and Clickhouse databases are not allowed to upload files to it so we need to stop showing such options to the users. This needs to happens for new DBs and existing ones. Also include Clickhouse DB as not allowed for data upload.

nytai commented on superset/views/database/forms.py:
```diff
@@ -39,6 +40,12 @@
+
+class DatabasesNotAllowedToFileUpload(str, Enum):
+    GSHEETS = "gsheets"
+    CLICKHOUSE = "clickhousedb"
```
Rather than maintaining a separate enum for this, why not define this as a property on the db_engine_spec?

Antonio-RiveroMartnez replied:
Hey @nytai! I just updated the PR with such property in our `base` and each engine now. Thanks for the suggestion.

rusackas replied:
@nytai let us know if you want us to recruit additional folks for a review, or if you have time to take a look. Thanks!

nytai replied:
I think having someone with more experience with these per-db configs, like @villebro or @john-bodley, have a look would be good.

nytai replied:
@Antonio-RiveroMartnez Thanks for making that change, multiple places in the frontend where you're checking 
`db?.engine !== Engines.GSheet && db?.engine !== Engines.ClickHouse`. This makes adding a new engine that does not support file upload a pain, since you have to update a bunch of files across both backend and frontend code. Ideally one could just specify in the db_engine_spec  `allows_file_upload=False` and that's it. So maybe adding a new api that you could call from the db modal to check is a supplied connection string or config supports file uploads would be a better approach?

villebro replied:
I agree with @nytai here - we should place this flag in the db engine specs, and the UI should accommodate accordingly. I may be wrong, but I believe we're already sending some db engine spec metadata in bootstrap data. Maybe we could just add `allows_file_upload` to `BaseEngineSpec`, override it in the GSheets and ClickHouse specs, and then make sure the flag is made available in bootstrap data? If we're not passing any info about db engine specs, now would probably be a good time to do so.

Antonio-RiveroMartnez replied:
Ok @nytai @villebro , I updated the PR, we were sending some `extra` information in the response but that might is editable by the user so I decided to include a new `engine_information` that currently only has `allows_file_upload` flag. Now, There is no need for updating multiple places but instead just setting such flags as true to prevent file uploading.

Antonio-RiveroMartnez committed:
GSheets and Clickhouse Databases:

- Add a new property in db_engine_spec so we can tell whether a db engine support file upload. It is True in by default.
- Stop using an Enum and instead make the method that populates the dropdown in our form use the new property for the engine of  a given database

Antonio-RiveroMartnez committed:
Databases:

- The text input Schemas allowed for File upload renders conditionally based on whther the user selects Allow file upload.
- Fix margins for checboxes and new location of the text input
- Update the wording for Allow file upload and text input for schemas
- Remove tooltip from Allow file upload checkbox
- Update tests

yousoph commented:
/testenv up

Antonio-RiveroMartnez committed:
Databases:

        - Include a new engine_information property in the response got our GET apis (single and list). Currently inside that property we send the allows_file_upload config value for the specific engine but it can added more information later.
        - Instead of manually checking if engine is GSheets or Clickhouse, use the new property so when new engines are added we dont need to change multiple file but just the property for it.
        - Update our tests and openapi spec to consider the new property

yousoph commented:
/testenv up

AAfghahi commented on superset/db_engine_specs/base.py:
```diff
@@ -1594,6 +1598,17 @@ def get_impersonation_key(cls, user: Optional[User]) -> Any:
+        :returns: Dict with properties of our class like allows_file_upload
+        """
+        return {
+            "allows_file_upload": cls.allows_file_upload,
```
how is this different from the current colum `allow_file_upload`? If they are very different, then should we think about changing the name of this? 

`allows_file_upload` and `allow_file_upload` being different columns that serve different purposes is going to be way too confusing.

Antonio-RiveroMartnez replied:
The `allow_file_upload` is in the `Database` model and it's set per each user's DB. While the `allows_file_upload` is set in the `DB Engine Spec` and cannot be set by the user. So, even when a `Database` has its flag set to true (think of existing user's DB), if its `DB Engine Spec` doesn't support it, we prevent file uploading to it.

AAfghahi replied:
ok that makes sense! 

Could we name it `supports_file_upload` or something along the like? Having two columns only separated by a singular/plural shift is going to be confusing.

Antonio-RiveroMartnez replied:
Yeah, totally! Let me make the change and ILYK. Thanks for the input @AAfghahi

AAfghahi replied:
of course! Otherwise this LGTM!

Antonio-RiveroMartnez replied:
PR Updated with the new `supports_file_upload` @AAfghahi  :)

Antonio-RiveroMartnez committed:
Databases:

- Rename allows_file_upload to supports_file_upload so we avoid confusion from the DB property

Antonio-RiveroMartnez committed:
Databases:

- Update our new RightMenu tests to include the environmentTag required prop

AAfghahi commented on superset-frontend/src/views/components/RightMenu.tsx:
```diff
@@ -118,8 +118,8 @@ const RightMenu = ({
   } = useSelector<any, ExtentionConfigs>(state => state.common.conf);
-  const [showModal, setShowModal] = useState<boolean>(false);
-  const [engine, setEngine] = useState<string>('');
+  const [showModal, setShowModal] = React.useState<boolean>(false);
```
What is the reason for using this versus importing in `useState` and invoking it normally?

Antonio-RiveroMartnez replied:
So we can `spyOn` during our tests, e.g. `const useStateMock = jest.spyOn(React, 'useState');` https://github.com/apache/superset/pull/21065/files#diff-6a8138bf7b3a4c95d31569dfe4497ceb65ae07faddc47fb944a611ec7fb19776R86

AAfghahi replied:
If you look at the line above that for useSelector, could you follow that method instead and import useState?

AAfghahi replied:
But that test and file still imports useState and uses it instead of doing `React.useState`

Antonio-RiveroMartnez replied:
Indeed but the usage of the spied hook is different which might be affecting why these tests cannot do it that way, either way I'll give a spin to see if it works that way, if not possible we need to decide whether we move forward with it as is or not 🤔 I'll keep you posted @AAfghahi , thanks for the input 🎉
